libqtile/notify.py

```diff
@@ -22,63 +23,55 @@
-        def __init__(self, manager):
-            bus = dbus.SessionBus()
-            bus.request_name(BUS_NAME)
-            bus_name = service.BusName(BUS_NAME, bus=bus)
-            service.Object.__init__(self, bus_name, SERVICE_PATH)
-            self.manager = manager
-            self._capabilities = {'body'}
-
-        @service.method(BUS_NAME, in_signature='', out_signature='as')
-        def GetCapabilities(self):  # noqa: N802
-            return list(self._capabilities)
-
-        def register_capabilities(self, capabilities):
-            if isinstance(capabilities, str):
-                self._capabilities.add(capabilities)
-            elif isinstance(capabilities, (tuple, list, set)):
-                self._capabilities.update(set(capabilities))
-
-        @service.method(BUS_NAME, in_signature='susssasa{sv}i', out_signature='u')
-        def Notify(self, app_name, replaces_id, app_icon, summary,  # noqa: N802
-                   body, actions, hints, timeout):
-            notif = Notification(
-                summary, body, timeout, hints, app_name, replaces_id, app_icon, actions
-            )
-            return self.manager.add(notif)
 
-        @service.method(BUS_NAME, in_signature='u', out_signature='')
-        def CloseNotification(self, id):  # noqa: N802
-            pass
+class NotificationService(ServiceInterface):
+    def __init__(self, manager):
+        super().__init__(BUS_NAME)
+        self.manager = manager
+        self._capabilities = {'body'}
+
+    @method()
+    def GetCapabilities(self) -> 'as':  # noqa: N802, F722
+        return list(self._capabilities)
+
+    def register_capabilities(self, capabilities):
+        if isinstance(capabilities, str):
+            self._capabilities.add(capabilities)
+        elif isinstance(capabilities, (tuple, list, set)):
+            self._capabilities.update(set(capabilities))
+
+    @method()
+    def Notify(self,  # noqa: N802, F722
+               app_name: 's', replaces_id: 'u',  # noqa: F821
+               app_icon: 's', summary: 's',  # noqa: F821
+               body: 's', actions: 'as',  # noqa: F821
+               hints: 'a{sv}', timeout: 'i') -> 'u':  # noqa: F821
+        notif = Notification(
+            summary, body, timeout, hints, app_name, replaces_id, app_icon, actions
+        )
+        return self.manager.add(notif)
+
+    @method()
+    def CloseNotification(self, id: 'u'):  # noqa: N802, F821
+        pass
 
-        @service.signal(BUS_NAME, signature='uu')
-        def NotificationClosed(self, id_in, reason_in):  # noqa: N802
-            pass
+    @signal()
+    def NotificationClosed(self, id_in, reason_in):  # noqa: N802
+        pass
 
-        @service.method(BUS_NAME, in_signature='', out_signature='ssss')
-        def GetServerInformation(self):  # noqa: N802
-            return ("qtile-notify-daemon", "qtile", "1.0", "1")
+    @method()
+    def GetServerInformation(self) -> 'ssss':  # noqa: N802, F821
```

Comment:
What is `'ssss'`? :D

Reply:
It's the output signature for that dbus method. `ssss` means that the method returns 4 strings (being `"qtile-notify-daemon"`, `"qtile"`, `"1.0"` and `"1"`).

While dbus_next's use of python's type hinting for the purposes of defining the signatures is pretty neat/clever. It does cause problems with mypy's type checking. Hence the need for extra `noqa`.